nixos/modules/services/development/zammad.nix

```diff
@@ -0,0 +1,217 @@
+      enable = lib.mkEnableOption "Zammad, a web-based, open source user support/ticketing solution.";
+
+      package = lib.mkOption {
+        type = lib.types.package;
+        default = pkgs.zammad;
+        defaultText = "pkgs.zammad";
+        description = "Zammad package to use.";
+      };
+
+      dataDir = lib.mkOption {
+        type = lib.types.path;
+        default = "/var/lib/zammad";
+        description = ''
+          Path to a folder that will contain Zammad working directory.
+        '';
+      };
+
+      host = lib.mkOption {
+        type = lib.types.str;
+        default = "127.0.0.1";
+        example = "192.168.23.42";
+        description = "Host address.";
+      };
+
+      port = lib.mkOption {
+        type = lib.types.int;
+        default = 3000;
+        description = "Web service port.";
+      };
+
+      websocketPort = lib.mkOption {
+        type = lib.types.int;
+        default = 6042;
+        description = "Websocket service port.";
+      };
+
+      dbName = lib.mkOption {
+        type = lib.types.str;
+        default = "zammad";
+        description = "The name of the database to use.";
+      };
+
+      dbUsername = lib.mkOption {
+        type = lib.types.str;
+        default = "zammad";
+        description = "The username to use to connect to the database.";
+      };
+
+      secretsFile = lib.mkOption {
+        type = lib.types.nullOr lib.types.path;
+        default = null;
+        description = ''
+          Path of a file containing secrets the format of EnvironmentFile as
+          described by systemd.exec(5). You must to define:
+            - PGPASSWORD
+            - SECRET_KEY_BASE
+          SECRET_KEY_BASE can be generated using:
+            ruby -e "require 'securerandom'; puts SecureRandom.hex(64)"
+        '';
+      };
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+
+    networking.firewall.allowedTCPPorts = [
+      config.services.zammad.port
+      config.services.zammad.websocketPort
+    ];
+
+    users.users.zammad = {
+      isSystemUser = true;
+      home = cfg.dataDir;
+      group = "zammad";
+    };
+
+    users.groups.zammad = {};
+
+    services.postgresql = {
+      enable = true;
+      ensureDatabases = [ cfg.dbName ];
+      ensureUsers = [
+        {
+          name = cfg.dbUsername;
+          ensurePermissions."DATABASE ${cfg.dbName}" = "ALL PRIVILEGES";
+        }
+      ];
+    };
+
+    systemd.services.zammad-setup = {
+      after = [ "postgresql.service" ];
+      requires = [ "postgresql.service" ];
+      serviceConfig.EnvironmentFile = cfg.secretsFile;
+      script = ''
+        echo "Setting password for database '${cfg.dbName}' and user '${cfg.dbUsername}'"
+        ${pkgs.utillinux}/bin/runuser -u ${config.services.postgresql.superUser} -- \
+          ${config.services.postgresql.package}/bin/psql \
+            -c "ALTER ROLE ${cfg.dbUsername} WITH PASSWORD '$PGPASSWORD'"
+        mkdir -p ${cfg.dataDir}
```

Comment:
systemd will do that automatically is part of `zammad.service` setup since `StateDirectory` is set